gruntwork-io/cloud-nuke
# Add config option for complex matching against s3 buckets

Addresses https://github.com/gruntwork-io/cloud-nuke/issues/108

Purpose
-------
Allow terminating AWS resources via more complex matching rules via a config file. 

Scope
------
This PR is scoped to support nuking s3 buckets based on matching by name.

CLI changes
-------------
I propose we allow something like this:

```sh
cloud-nuke aws --config path/to/file.yaml
```

Yaml config format
-----------------
The structure of the config should reflect the way AWS describes resources. S3 buckets have names and regions, so for the scope of this project:

```yaml
s3:
  include:
    names_regex:
      - ^alb-alb-.*-access-logs$
      - ...
```

The proposed format at a meta level is this:
```
[RESOURCE_TYPE]:
  [include | exclude]:
    [FIELD_PLURAL][_regex]:
      - [ list of matchers ]
         OR
     [key]: [value]
```
where
  * `resource_type` can be `s3`, `iam_role`, etc. _We are punting on supporting `region` for now._
  * `include` has rules that a resource must match to be nuked
  * `exclude` has rules that a resource must not match to be nuked
  * matching key looks like:
    * `tags_regex`: would accept a list of regular expressions
    * `names_regex`: would accept a list of regular expressions
    * `regions` (assumes plaintext, exact matching)
  * for resources like `ec2_instance`, we need to support key-value maps
    * e.g. `Name: ^test_.*`

Looking Ahead
---------------
<details><summary>[expand]</summary>

Eventually, we might support something like:

```yaml
ec2_instance:
  include:
    tags_regex:
      Name: ^test_.*

iam_role:
  include:
    names_regex:
      - ^test_.*
      - ^cloud-nuke-TestListEksClusters-.*

s3:
  include:
    names_regex:
      - ^alb-alb-.*-access-logs$
```

The following would nuke _only_ s3 buckets matching the `names_regex` but that are not in `us-west-1`, if called with `cloud-nuke aws --resource-type s3 --config path/to/config.yaml`:

```yaml
s3:
  include:
    names_regex:
      - ^alb-alb-.*-access-logs$
  exclude:
    regions:
      - us-west-1
```

</details>


How to test locally
-------------------

<details><summary>[expand]</summary>

1. Clone this repo in your regular development directory (no need to use $GO_DIR)
1. Log into nuclear wasteland (if you need access we need to grant it to you)
1. Run some commands like:
    * `go run main.go aws --resource-type s3 --config config/mocks/s3_cleanup.yaml`
    * try the command with/out `resource-type` option
    * try the command with/out `config` option
1. Run tests against sandbox environment, locally:
    * ```sh
      cd {repo}/aws
      houston exec sandbox -- go test -v -run TestFilterS3
      ```

</details>



Summary of trial project
--------------------------

<details><summary>[expand]</summary>

Notably completed today:
- [x] Add `--config path/to/file.yaml` support to the command line.
- [x] Add support for a simple config yaml including s3 buckets by name, using regexp.
- [x] Confirm via manual testing that these following commands work:
    - `go run main.go aws`: nuke all s3 buckets
    - `go run main.go aws --config ./config/mocks/s3_include_names.yaml`: nuke only s3 buckets in all regions that match any of the regexps.
    - `go run main.go aws --config ./config/mocks/s3_include_names.yaml --region us-east-1`: nuke only s3 buckets in us-east-1 that match any of the regexps.

Notably not done:
- [x] Add test coverage in `config/config_test.go`
- [ ] Support other kinds of keys in the yaml config, like `exclude_names_regex`, `exclude_names`, and `include_names`.

Of course this is still far from complete, but might set some ground work for the future. There are still a few TODOs in the code that signal what might be worth checking. One caveat to mention: I haven't written `go` in years and was never an expert, so some of the conventions might be lost on me. (But that doesn't make it any less fun for me!)
</details>
